src/hsm.cpp

```diff
@@ -104,36 +117,60 @@ openssl::SSL_EVP_PKEY_Ptr HsmEngine::generateKey(const RSASpec &spec,
                                                  const std::string &keyLabel,
                                                  const std::vector<uint8_t> &keyID)
 {
-    std::string keyIDHexString = utility::toHex(keyID);
-    PKCS11_RSA_KGEN pkcs11RSASpec;
-    pkcs11RSASpec.bits = spec.numberOfBits();
-    PKCS11_KGEN_ATTRS pkcs11RSAKeygen;
-    pkcs11RSAKeygen.type = EVP_PKEY_RSA;
-    pkcs11RSAKeygen.kgen.rsa = &pkcs11RSASpec;
-    pkcs11RSAKeygen.key_id = keyIDHexString.c_str();
-    pkcs11RSAKeygen.token_label = _tokenLabel.c_str();
-    pkcs11RSAKeygen.key_label = keyLabel.c_str();
+    try {
+        // We need to make sure that we don't have 2 keys with the same ID.
+        // For that we need to pass empty keyLabel. Otherwise libp11 tries to find
+        // a key with exact keyLabel/keyID combination. This means that libp11 might
+        // not recognize that the key with the same ID is already there.
+        loadPrivateKey("", keyID);
+    } catch (const MoCOCrWException &e) {
+        if (e.what() == std::string("Unable to load private key. Private key not found!")) {
+            std::string keyIDHexString = utility::toHex(keyID);
```

Comment:
It feels a bit weird that all this code is now in the catch blog. I'm also not a 100% sure if this is good. What happens if this code throws again (while its still handling an exception?
Can we move this out of the catch block?